src/v/redpanda/admin_server.h

```diff
@@ -147,27 +148,32 @@ class admin_server {
           [this, handler](std::unique_ptr<ss::http::request> req)
             -> ss::future<ss::json::json_return_type> {
               auto auth_state = apply_auth<required_auth>(*req);
-
-              // Note: a request is only logged if it does not throw
-              // from authenticate().
-              log_request(*req, auth_state);
-
-              // Intercept exceptions
-              const auto url = req->get_url();
-              if constexpr (peek_auth) {
-                  return ss::futurize_invoke(
-                           handler, std::move(req), auth_state)
-                    .handle_exception(
-                      exception_intercepter<
-                        decltype(handler(std::move(req), auth_state).get0())>(
-                        url, auth_state));
-
-              } else {
-                  return ss::futurize_invoke(handler, std::move(req))
-                    .handle_exception(exception_intercepter<
-                                      decltype(handler(std::move(req)).get0())>(
-                      url, auth_state));
-              }
+              return ss::do_with(
+                std::move(auth_state),
+                [this, handler, req = std::move(req)](
+                  const auto& auth_state) mutable {
+                    // Note: a request is only logged if it does not throw
+                    // from authenticate().
+                    log_request(*req, auth_state);
+
+                    // Intercept exceptions
+                    const auto url = req->get_url();
+                    if constexpr (peek_auth) {
+                        return ss::futurize_invoke(
+                                 handler, std::move(req), auth_state)
+                          .handle_exception(
+                            exception_intercepter<
+                              decltype(handler(std::move(req), auth_state)
+                                         .get0())>(url, auth_state));
+
+                    } else {
+                        return ss::futurize_invoke(handler, std::move(req))
+                          .handle_exception(
+                            exception_intercepter<
+                              decltype(handler(std::move(req)).get0())>(
+                              url, auth_state));
+                    }
+                });
```

Comment:
nice catch.  Pretty much a perfect example of this: https://docs.seastar.io/master/split/8.html#keeping-ownership-at-the-caller